### ISSUE TITLE
[BUG] failed to ready HDA controller gctl 0x0 when suspend-resume-with-capture on MTL platforms

### ISSUE THREAD

**Issue by miRoox** [thesofproject/linux#4077]
**Describe the bug**
CI observed the error "failed to ready HDA controller gctl 0x0" when suspend-resume on MTLP_RVP_SDW.

**To Reproduce**
```
TPLG=/lib/firmware/intel/sof-ace-tplg/sof-mtl-rt711-4ch.tplg MODEL=MTLP_RVP_SDW ~/sof-test/test-case/check-suspend-resume-with-audio.sh -l 5 -m capture
```

**Reproduction Rate**
About 80%

**Environment**
1) Branch name and commit hash of the 2 repositories: sof (firmware/topology) and linux (kernel driver).
    * Kernel: https://github.com/thesofproject/linux/commit/321028577de9
    * SOF: https://github.com/thesofproject/sof/commit/4a9a99aeb24a
2) Name of the topology file
    * Topology: sof-ace-tplg/sof-mtl-rt711-4ch.tplg
3) Name of the platform(s) on which the bug is observed.
    * Platform: MTLP_RVP_SDW

**Screenshots or console output**
**dmesg**
```
[ 1457.032024] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed to ready HDA controller gctl 0x0
[ 1457.032029] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed to exit HDA controller reset
[ 1457.032063] kernel: snd_sof:snd_sof_pci_update_bits_unlocked: sof-audio-pci-intel-mtl 0000:00:1f.3: Debug PCIR: 00000000 at  00000048
[ 1457.032071] kernel: snd_sof:snd_sof_pci_update_bits_unlocked: sof-audio-pci-intel-mtl 0000:00:1f.3: Debug PCIW: 00000040 at  00000048
[ 1457.032075] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed to start controller after resume
[ 1457.032079] kernel: snd_sof_intel_hda_common:hda_dsp_state_log: sof-audio-pci-intel-mtl 0000:00:1f.3: Current DSP power state: D0I0
[ 1457.032083] kernel: snd_sof:sof_set_fw_state: sof-audio-pci-intel-mtl 0000:00:1f.3: fw_state change: 0 -> 2
[ 1457.032086] kernel: snd_sof:snd_sof_load_firmware: sof-audio-pci-intel-mtl 0000:00:1f.3: loading firmware
[ 1457.032089] kernel: snd_sof:sof_set_fw_state: sof-audio-pci-intel-mtl 0000:00:1f.3: fw_state change: 2 -> 3
[ 1457.083593] kernel: snd_sof_intel_hda_common:mtl_dsp_pre_fw_run: sof-audio-pci-intel-mtl 0000:00:1f.3: FW Poll Status: reg[0x1000]=0x0 timedout
[ 1457.083598] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: failed to enable DSP subsystem
[ 1457.083601] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: failed pre fw run op
[ 1457.083604] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed to boot DSP firmware after resume -110
[ 1457.083607] kernel: snd_sof:sof_set_fw_state: sof-audio-pci-intel-mtl 0000:00:1f.3: fw_state change: 3 -> 4
[ 1457.083611] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: PM: dpm_run_callback(): pci_pm_resume+0x0/0x80 returns -110
[ 1457.083626] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: PM: failed to resume async: error -110
[ 1457.284069] kernel: soundwire_intel:intel_link_power_up: soundwire_intel soundwire_intel.link.0: powering up all links
[ 1457.284073] kernel: soundwire_intel:intel_link_power_up: soundwire_intel soundwire_intel.link.0: first link up, programming SYNCPRD
[ 1457.285415] kernel: soundwire_intel soundwire_intel.link.0: Failed to power up link: -11
[ 1457.285420] kernel: soundwire_intel soundwire_intel.link.0: intel_resume failed: -11
[ 1457.285423] kernel: soundwire_intel soundwire_intel.link.0: PM: dpm_run_callback(): pm_generic_resume+0x0/0x30 returns -11
[ 1457.285435] kernel: soundwire_intel soundwire_intel.link.0: PM: failed to resume: error -11
```

[full dmesg.txt](https://github.com/thesofproject/sof/files/10162960/dmesg.txt)
[mtrace.txt](https://github.com/thesofproject/sof/files/10162958/mtrace.txt)

**Comment by miRoox** [comment]
This issue also happened with IPC4 Tester, transfer to the kernel side.

**Comment by plbossart** [comment]
@miRoox can you try to see if https://github.com/thesofproject/linux/pull/4082 helps?

**Comment by ranj063** [comment]
@plbossart I do not think this has anything to do with the sleep. We had a similar issue on TGL and it was related to the DSP clock https://github.com/thesofproject/linux/issues/2571

**Comment by plbossart** [comment]
it seems different @ranj063. The TGL issue was on probe, before reading the caps. 

This is on resume. This is the second case where we reset the link.

**Comment by ranj063** [comment]
In any case @plbossart D3 implementation on MTL is still  incomplete in the FW, so I think we should wait until after it is done to continue debugging this issue

**Comment by plbossart** [comment]
@ranj063 I thought about this, but it also happens with the reference firmware so there's something else at play.

**Comment by ranj063** [comment]
> @ranj063 I thought about this, but it also happens with the reference firmware so there's something else at play.

Oh I missed that part! It is interesting though that this problem only manifest with suspend/resume with audio

**Comment by miRoox** [comment]
> @miRoox can you try to see if #4082 helps?

After testing multiple times with #4082, I can only see https://github.com/thesofproject/sof/issues/6679 when suspend-resume, so I think it helps this issue.

**Comment by plbossart** [comment]
@miRoox have you tested with the reference firmware? We need an apples-to-apples comparison and the SOF part is still too new to conclude on what hardware timings need to be.

**Comment by miRoox** [comment]
> have you tested with the reference firmware?

the reference firmware has many different errors, and this failure has less reproduction rate.

**Comment by keqiaozhang** [comment]
Confirmed that this issue can be fixed by https://github.com/thesofproject/linux/pull/4082

**Comment by mengdonglin** [comment]
More stressful test shows:

- https://github.com/thesofproject/linux/pull/4082 can greatly lower the reproduce rate of this issue, but cannot totally avoid it.
- This issue seems to only happen after DSP enters in D0ix to do deep buffer playback.  If  we prevent DSP from entering D0ix, this issue will NOT happen, proved by the test below: 
   This issue CANNOT be reproduced If we skip deep buffer pipeline in test with runtime PM enabled. In this case, ADSP only switch between D0 (active playback/capture of other stream)  and D3hot (suspended);
  This issue also CANNOT be reproduced if keep deep buffer pipeline in test with runtime PM disabled. In this case, ADSP will stay in D0 even for doing deep buffer playback only. And enter D3 when idle.

**Comment by miRoox** [comment]
> * This issue also CANNOT be reproduced if keep deep buffer pipeline in test with runtime PM disabled.

We can reproduce this issue when suspend-resume if keep deep buffer pipeline in test with runtime PM disabled

**Comment by ranj063** [comment]
@RanderWang @mengdonglin can we file a new issue for the deep buffer playback scenario? I think the original problem mentioned in this issue and the problem now have diverged.

In any case, when the DSP is in D0i3 and you do a system suspend @RanderWang can we please confirm if the DSP exits D0i3 to enter D0i0 and then suspends to D3?

**Comment by plbossart** [comment]
Agree with @ranj063, I can't follow the various configurations and what works or doesn't. Let's file one issue per failed test case, not lump them all in the same issue of a failed reset.

**Comment by miRoox** [comment]
> can we file a new issue for the deep buffer playback scenario?

We have https://github.com/thesofproject/sof/issues/6938 for that scenario, maybe we should transfer it to Linux repository?

**Comment by RanderWang** [comment]
> In any case, when the DSP is in D0i3 and you do a system suspend @RanderWang can we please confirm if the DSP exits D0i3 to enter D0i0 and then suspends to D3?

Thanks!  If there is a d0i3 compatible stream like deep buffer, we never go to d3, only d0i3. confirmed by log.

**Comment by RanderWang** [comment]
please check https://github.com/thesofproject/sof/issues/6938. Before runtime suspend ,  dsp is in d0i0, not d0i3.

**Comment by mengdonglin** [comment]
This issue disappeared on MTLP-RVP-HDA in CI daily test after #4082 merged on Jan 30. 

But we can still reproduce this issue with more stressful testing with ADSP previously enters D0i3.  Unfortunately, applying latest PMC code cannot solve this issue: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6aa5207838ccf79879f212034435a5606f890eea (platform/x86: intel/pmc/core: Add Meteor Lake mobile support)

**Comment by mengdonglin** [comment]
This issue cannot be reproduced after #4173 got merged. Closing.
